File: 127.word-ladder.py
#
# @lc app=leetcode id=127 lang=python3
#
# [127] Word Ladder
#

# @lc code=start


from collections import deque
from string import ascii_lowercase


class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        wordSet = set(wordList)
        if endWord not in wordSet:
            return 0
        bq, eq = deque([(1, beginWord)]), deque([(1, endWord)])
        bpos, epos = {beginWord: 1}, {endWord: 1}

        while bq and eq:
            for bdist, st, c, en in (
                (bdist, w[:i], c, w[i + 1 :])
                for bdist, w in (bq.popleft() for _ in range(len(bq)))
                for i, c in enumerate(w)
            ):
                for nw in (f"{st}{l}{en}" for l in ascii_lowercase if l != c):
                    if edist := epos.get(nw):  # BFS from both ends converge
                        return bdist + edist
                    if nw in wordSet:
                        bpos[nw] = bdist + 1
                        bq.append((bdist + 1, nw))
                        wordSet.remove(nw)
            if len(bq) > len(eq):  # swap to minimize BFS tree
                bq, bpos, eq, epos = eq, epos, bq, bpos
        return 0


# @lc code=end

File: 127.word-ladder.rs
/*
 * @lc app=leetcode id=127 lang=rust
 *
 * [127] Word Ladder
 */

// @lc code=start
use std::{
    collections::{HashMap, HashSet, VecDeque},
    mem::swap,
};

impl Solution {
    pub fn ladder_length(begin_word: String, end_word: String, word_list: Vec<String>) -> i32 {
        let mut word_set: HashSet<String> = word_list.into_iter().collect();
        if !word_set.contains(&end_word) {
            return 0;
        }
        let mut bq = VecDeque::from([(1, begin_word.clone())]);
        let mut eq = VecDeque::from([(1, end_word.clone())]);
        let mut bpos = HashMap::from([(begin_word, 1)]);
        let mut epos = HashMap::from([(end_word, 1)]);

        while !bq.is_empty() && !eq.is_empty() {
            for _ in 0..bq.len() {
                let (dist, word) = bq.pop_front().unwrap();
                for (i, c) in word.chars().enumerate() {
                    let (begin, end) = (&word[..i], &word[i + 1..]);
                    for l in ('a'..='z').filter(|&l| l != c) {
                        let nw = format!("{}{}{}", begin, l, end);
                        if let Some(&d) = epos.get(&nw) {
                            return dist + d;
                        }
                        if word_set.contains(&nw){
                            word_set.remove(&nw);
                            bpos.insert(nw.clone(), dist + 1);
                            bq.push_back((dist + 1, nw));
                        }
                    }
                }
            }
            if bq.len() > eq.len() {
                swap(&mut bq, &mut eq);
                swap(&mut bpos, &mut epos);
            }
        }
        0
    }
}
// @lc code=end
